### DISCUSSION TITLE
Artisan::call() make infinit loop

### DISCUSSION THREAD

**Discussion by moradi-morteza** [laravel/framework#38459] [Q&A]
- Laravel Version: 8.48.0
- PHP Version: 7.4.2
- Database Driver & Version:

### Description:
i have a function contain some Artisan Command when i run this function from a controller , **the function run many time and not stop** i think the problem is from Artisan::call('migrate') probably.

```
function change_database(){
    Artisan::call('env:set APP_DATABASE '.$db_name);
    Artisan::call('config:cache');
    Artisan::call('vendor:publish --tag=migrations.modules');
    Artisan::call('migrate', ['--database' => 'base', '--path' => '/database/migrations/base']);
    Artisan::call('migrate', ['--database' => 'data', '--path' => '/database/migrations/data']);
}
```
also i create a custom command and put function inside it, then run command from terminal, **its run without problem.**
**where is the problem?**

**Comment by challgren** [comment]
I have this same issue with version v7.30.4. If I call the command using exec() it doesn't loop. I'm calling it from a job.

**Reply by brunogaspar** [reply]
Most likely you need to pass the `--force` option:

```php
Artisan::call('migrate', [
    '--force' => true
]);
```

This is specially required when you're on production.

If it still hangs, then your code definitely has a problem.

**Reply by moradi-morteza** [reply]
thanks, I'm checking this out

**Reply by challgren** [reply]
I will say my command is not a built in Laravel command, its a custom command. So I don't think the issue is related to calling migration. The real issue is in `Artisan::call()`

**Reply by brunogaspar** [reply]
Without knowing what you are doing exactly, is quite difficult to help.

**Reply by challgren** [reply]
```
<?php

namespace App\Jobs;

use App\Exceptions\ReservedJobException;
use Exception;
use Illuminate\Support\Carbon;
use Illuminate\Support\Facades\Artisan;
use Illuminate\Support\Facades\DB;

class ProcessInsightsPosFile extends FileJob
{
    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        if ($this->attachment->reserved_at) {
            throw new ReservedJobException($this->attachment);
        }

        $this->attachment->reserved_at = Carbon::now();
        $this->attachment->save();
        $output = [];
        $returnValue = 0;
        $command = $this->getCommand(
            sprintf(
                'process_insight_not_in_pos.py %s %s %s %s',
                $this->attachment->getS3Path(),
                $this->attachment->audit_id,
                $this->attachment->user_id,
                $this->attachment->id
            )
        );

        exec($command, $output, $returnValue);

        if ($returnValue != 0) {
            $this->attachment->reserved_at = null;
            $this->attachment->save();
            throw new Exception(implode("\n", [$command] + $output));
        }

        Artisan::call('app:process-debookings', [
            'audit_id' => $this->attachment->audit_id,
            'type' => 'insight_not_in_pos'
        ]);

        $this->updateResults(DB::raw('TRIM(`audit_insights_not_in_pos`.`serial_number`)'), 'audit_data_partners.serial_number');
        $this->updateResults(DB::raw('TRIM(`audit_insights_not_in_pos`.`serial_number`)'), 'audit_data_partners.serial_number_alt');
        $this->updateResults('audit_insights_not_in_pos.serial_number_alt', 'audit_data_partners.serial_number');

        Artisan::call('app:audit-stats', ['audit_id' => $this->attachment->audit_id, 'action' => 'insight-pos-file']);

        $this->attachment->completed_at = Carbon::now();
        $this->attachment->save();
    }
```

The `Artisan::call('app:process-debookings', ['audit_id' => $this->attachment->audit_id, 'type' => 'insight_not_in_pos']);` is the problem in my function handle() If I Log::info('Processing debookings for ' . $audit_id . ' type ' . $type); it results in a never ending loop until the container is killed due to too many connections. At first I thought I was going crazy because the queue process would get killed because the DB would abort due to too many connections. Even Xdebug alerted me about recursion. I then started looking for others having the issue and came across this discussion that is related to 8.x but it appears its an issue in 7.x and 8.x. I did just upgrade to the latest 8.x and I will test on Monday to see if I still have the issue.

**Reply by brunogaspar** [reply]
So, without that Artisan call, does the job succeed?

If so, i would try to debug that command and see if you receive all the data as expected, from the (job => command), then try to disable a few lines within it and try it again. It might be a bit tedious, but it's part of debugging lol, however it seems something within your code is causing the issue.

I don't have anything that triggers Artisan commands within jobs, but i do have a simple project that triggers one Artisan command within a controller, so it's not async per se, but it does some tricky things within it and it kinda works as expected.

**Reply by challgren** [reply]
Yes if I call it via exec() it works fine. If I called it via Artisan::call it does an infinite loop.

**Reply by challgren** [reply]
I upgraded to Laravel 8.69.0 and this issue still exists. Also this issue exists in multiple Jobs/Commands so its not related to a single Job or Command

**Reply by brunogaspar** [reply]
Then i recommend you to create a repository that reproduces this problem, so we can try to figure out what is the cause and determine if it's a problem on your code or within the framework.
